client/executor/common/src/wasm_runtime.rs

```diff
@@ -19,6 +19,42 @@
 use crate::error::Error;
 use sp_wasm_interface::Value;
 
+/// Call site.
+///
+/// Contains variants on how to resolve wasm function that will be invoked.
+pub enum InvokeMethod<'a> {
+	/// Call function exported with this name.
+	///
+	/// Located function should have (u32, u32) -> u64 signature.
+	Export(&'a str),
+	/// Call function by reference from table.
+	///
+	/// Located function should have (u32, u32) -> u64 signature.
+	Table(u32),
+	/// Call function by reference from table through a wrapper.
+	///
+	/// Invoked function (`dispatcher_ref`) function
+	/// should have (u32, u32, u32) -> u64 signature.
+	///
+	/// `func` will be passed to the invoked function as a first argument.
+	TableWithWrapper {
+		/// Wrapper for the call.
+		///
+		/// Function pointer, index into runtime exported table.
+		dispatcher_ref: u32,
+		/// Actual wasm function pointer (index into table) that should be invoked.
```

Comment:
Funnily, this is not necessarily a function index, but rather arbitrary designator. I.e. consider a dispatcher that is implemented as a switch statement instead of `call_indirect`ing the given function index.